File: prime.hs
generatelist    :: Integer -> [Integer]
generatelist n
    | n >= 2 = 2:[3,5..n]
    | n <  2 = []

multiples       :: Integer -> Integer -> [Integer]
multiples n max = [n, 2*n..max]

remove_from_list :: [Integer] -> [Integer] -> [Integer] -> [Integer]
remove_from_list target removal_array rv
    | null target || null removal_array =
        rv ++ target
    | head target > head removal_array =
        remove_from_list target (tail removal_array) rv
    | head target == head removal_array =
        remove_from_list (tail target) (tail removal_array) rv
    | otherwise =
        remove_from_list (tail target) removal_array (rv ++ [head target])

findprimes :: Integer -> [Integer]
findprimes n =
    findprimes' n (generatelist n) []
    
findprimes' :: Integer -> [Integer] -> [Integer] -> [Integer]
findprimes' n candidates primes
    | null candidates = primes
    | otherwise =
        findprimes'
            n
            (remove_from_list
                candidates
                (multiples (head candidates) n)
                [])
            (primes ++ [head candidates])

File: prime.lua
-- prime number generator

function getprimes (n)
    if n < 2 then
        primes = {}
    else
        primes = {2}
        candidates = {}
        for i=3,n,2 do
            table.insert(candidates, i)
        end
        while candidates[1] ~= nil do
            table.insert(primes, candidates[1])
            remove_array = multiples(candidates[1], n)
            counter = 1
            while remove_array[1] ~= nil and candidates[counter] ~= nil do
                if remove_array[1] < candidates[counter] then
                    table.remove(remove_array, 1)
                elseif remove_array[1] == candidates[counter] then
                    table.remove(candidates, counter)
                else
                    counter = counter + 1
                end
            end
        end
    end
    return primes
end

function multiples (n, max)
    i = 1
    m = {}
    while i*n <= max do
        table.insert(m, i*n)
        i = i + 1
    end
    return m
end

primes = getprimes(100)
for i,v in ipairs(primes) do
    print(v)
end

